test/Streamly/Test/FileSystem/DirIO.hs

```diff
@@ -0,0 +1,143 @@
+-- |
+-- Module      : Streamly.Test.FileSystem.DirIO
+-- Copyright   : (c) 2019 Composewell Technologies
+-- License     : BSD-3-Clause
+-- Maintainer  : streamly@composewell.com
+-- Stability   : experimental
+-- Portability : GHC
+
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE QuasiQuotes #-}
+
+module Main (main) where
+
+--------------------------------------------------------------------------------
+-- Imports
+--------------------------------------------------------------------------------
+
+import Data.Word (Word8)
+import GHC.IO.Encoding (setLocaleEncoding, utf8)
+import Streamly.Unicode.String (str)
+import System.Process (readCreateProcess, shell)
+#if !defined(mingw32_HOST_OS) && !defined(__MINGW32__)
+import Streamly.Data.Array (Array)
+#endif
+
+import qualified Streamly.Unicode.Stream as Unicode
+import qualified Streamly.Internal.Unicode.Stream as Unicode (lines)
+import qualified Streamly.Data.Stream.Prelude as Stream
+import qualified Streamly.Data.Fold as Fold
+import qualified Streamly.Data.StreamK as StreamK
+
+import Prelude hiding (last, length)
+import BenchTestLib.DirIO
+
+import Test.Hspec as H
+
+--------------------------------------------------------------------------------
+-- Main
+--------------------------------------------------------------------------------
+
+moduleName :: String
+moduleName = "FileSystem.DirIO"
+
+testCorrectness
+    :: Stream.Stream IO [Char] -> Stream.Stream IO Word8 -> Expectation
+testCorrectness strmBase lister = do
+    let strm =
+            StreamK.toStream
+                $ StreamK.sortBy compare
+                $ StreamK.fromStream
+                $ Unicode.lines Fold.toList
+                $ Unicode.decodeUtf8 lister
+    Stream.eqBy (==) strm strmBase `shouldReturn` True
+
+#if !defined(mingw32_HOST_OS) && !defined(__MINGW32__)
+testCorrectnessByteChunked
+    :: Stream.Stream IO [Char] -> Stream.Stream IO (Array Word8) -> Expectation
+testCorrectnessByteChunked strmBase lister = do
+    let strm =
+            StreamK.toStream
+                $ StreamK.sortBy compare
+                $ StreamK.fromStream
+                $ Unicode.lines Fold.toList
+                $ Unicode.decodeUtf8Chunks lister
+    Stream.eqBy (==) strm strmBase `shouldReturn` True
+#endif
+
+-- | List the current directory recursively
+main :: IO ()
+main = do
+    setLocaleEncoding utf8
+
+    let smallTree = "benchmark-tmp/dir-structure-small"
+        bigTree = "benchmark-tmp/dir-structure-big"
+    createDirStucture smallTree 2 3
+    createDirStucture bigTree 5 5
+
+    findResBig <- readCreateProcess (shell [str|find #{bigTree}|]) ""
+    findResSmall <- readCreateProcess (shell [str|find #{smallTree}|]) ""
```

Comment:
You can add a todo here, createDirStructure itself can return the created structure, so we do not have to depend on `find` and `process`.